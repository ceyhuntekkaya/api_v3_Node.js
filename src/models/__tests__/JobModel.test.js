const HarvestApiConnector = require('../../connectors/HarvestApiConnector')
const JobModel = require('../JobModel')
const JobObj = require('./__data__/jobObj.json')

const mockConnection = ({ userId }) => ({ method, endpoint, data }) => {
  const urlToPayload = {
    'jobs': JobObj,

    'jobs/123456/hiring_team': { success: true },

    'jobs/100001/hiring_team': { success: true },

    'jobs/123456': { userId, method, endpoint, data },

    'jobs/100001': JobObj,

    'custom_fields/job': [
      {
        id: 131029,
        field_type: 'job',
        priority: 2,
        value_type: 'single_select',
        private: false,
        required: true,
        require_approval: false,
        trigger_new_version: false,
        name: 'Employment Type',
        name_key: 'employment_type',
        active: true,
        custom_field_options: [
          {
            id: 10684619,
            name: 'Seasonal',
            priority: 0
          },
          {
            id: 11784837,
            name: 'Year Round',
            priority: 1
          },
          {
            id: 11220411,
            name: 'Part Time',
            priority: 2
          }
        ]
      }
    ],

    'users': {
      id: 1017652,
      first_name: 'first',
      last_name: 'last'
    },

    'users/597529/permissions/future_jobs?per_page=125': [
      {
        id: 361480,
        office_id: 27259,
        department_id: null,
        user_role_id: 17735
      }
    ],

    'users/597529/permissions/jobs?per_page=500': [
      {
        id: 46288285,
        job_id: 741762,
        user_role_id: 17735
      }
    ],

    'users/123/permissions/future_jobs': {
      id: '123',
      user_role_id: '566',
      office_id: '999'
    },

    'users?email=test@email.com': {
      id: 123456,
      name: 'test',
      first_name: 'first_test',
      last_name: 'last_test',
      updated_at: '2019-02-28T17:12:22.818Z',
      created_at: '2019-02-28T17:12:22.818Z',
      disabled: false,
      site_admin: false,
      emails: [
        'test@email.com'
      ],
      employee_id: null
    }
  }

  const payload = urlToPayload[endpoint]

  if (!payload) {
    throw new Error(`unmocked endpoint: '${endpoint}'`)
  }

  return Promise.resolve(payload)
}

describe('Job Model', () => {
  const harvest = HarvestApiConnector({ connection: mockConnection({ userId: 786605 }) })
  const model = JobModel({ harvest })

  it('can get custom fields', () => {
    return expect(model.getCustomFields())
      .resolves
      .toMatchSnapshot()
  })

  it('can save a new job', () => {
    Math.random = jest.fn(() => 1)

    const newJob = {
      userId: 786605,
      office: 27259,
      applicantGroup: 11382848,
      jobCategory: 11630500,
      jobTitle: 'Test Job -sw',
      employmentType: 10684619,
      yearsExperience: 11220424,
      submissionDeadline: '2018-04-30',
      baseCompensation: 50,
      baseCompensationMax: 80,
      paySchedule: 11220408,
      additionalIncome: 'additional income field',
      totalCompensation: 70,
      contactName: 'test user',
      contactTitle: 'test title',
      contactEmail: 'test@testmail.com',
      education: 11220428,
      applyVia: 11612190,
      website: 'https://www.test.com/jobs',
      overview: 'overview field',
      experienceRequired: '<ul><li>experience field</li></ul>',
      responsibilities:
        '<em>essential responsiblities field</em>benefits: <strong>benefits field</strong>',
      benefits: '<h1>benefits</h1>',
      termOfServiceAgreement: true
    }

    return expect(model.create(newJob))
      .resolves
      .toMatchSnapshot()
  })

  it('can get job permissions', () => {
    return expect(model.getJobPermissions('597529'))
      .resolves
      .toMatchSnapshot()
  })

  it('can create job permissions', async () => {
    return expect(model.createJobPermission(123,
      {
        user_role_id: '566',
        office_id: '999'
      }))
      .resolves
      .toMatchSnapshot()
  })

  it('can update job', () => {
    return expect(model.update(100001, { currentStep: 'Removed' }))
      .resolves
      .toMatchSnapshot()
  })

  it('can duplicate job', () => {
    return expect(model.duplicate(100001, 'job Title (Copy)'))
      .resolves
      .toMatchSnapshot()
  })

  it('can create user', () => {
    const payload = {
      first_name: 'firstName',
      last_name: 'lastName',
      primaryEmail: 'first@email.com'
    }
    return expect(model.createUser(payload))
      .resolves
      .toMatchSnapshot()
  })

  it('can find user by email', () => {
    const email = 'test@email.com'
    return expect(model.findUserByEmail(email))
      .resolves
      .toMatchSnapshot()
  })
})

const GREENHOUSE_ID = '123'

const EXPECTED_MY_JOBS = {
  [GREENHOUSE_ID]: [{ job_id: 334 }]
}

const ALGOLIA_JOBS = [{
  id: '334',
  jobId: 334,
  userId: '0',
  hiringManager: {
    id: '',
    firstName: '',
    lastName: '',
    employeeId: ''
  },
  applicantGroup: '',
  job_category: 'Seasonal',
  jobTitle: 'PGA Pro',
  status: 'open',
  employment_type: 'Hourly',
  years_experience: '2 years',
  submission_deadline: 1530158400,
  base_compensation: { value: 14 },
  base_compensation_max: { value: 55 },
  pay_schedule: '',
  additional_income: null,
  total_compensation: { value: 66 },
  job_contact_name: 'John Newkirk',
  job_contact_title: 'PGA Pro',
  job_contact_email: 'jnewkirk@pgahq.com',
  education_required: ["bachelor's"],
  apply_via: null,
  website: 'http://website.com/',
  overview: '',
  experience_required: '2 years',
  responsibilities: 'Golf professional',
  benefits: 'free rounds',
  tos: false,
  current_step: 'Completed',
  opened: 1529012518,
  created: 1529012518,
  facilities: [{
    greenhouse_id: 123,
    facility_name: 'PGAHQ'
  }]
}]

describe('Job Model', () => {
  test('getJobsPostedBy', () => {
    const algoliaMock = {
      findBatch: (index, jobIds) => {
        const results = ALGOLIA_JOBS.filter(j => jobIds.includes(j.id))
        return Promise.resolve(results)
      }
    }

    const harvestMock = {
      getMyJobs: (greenhouseId) => Promise.resolve(EXPECTED_MY_JOBS[greenhouseId])
    }

    const model = JobModel({
      algolia: algoliaMock,
      harvest: harvestMock
    })

    const member = { greenhouseId: GREENHOUSE_ID }

    return expect(model.getJobsPostedBy(member))
      .resolves
      .toMatchSnapshot()
  })
})
