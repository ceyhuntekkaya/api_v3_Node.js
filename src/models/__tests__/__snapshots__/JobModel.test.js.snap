// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Model can create job permissions 1`] = `
Object {
  "id": "123",
  "officeId": "999",
  "userRoleId": "566",
}
`;

exports[`Job Model can create user 1`] = `
Object {
  "first_name": "first",
  "id": 1017652,
  "last_name": "last",
}
`;

exports[`Job Model can duplicate job 1`] = `
Object {
  "additionalIncome": null,
  "applicantGroup": "PGA Professional and General Public",
  "applyVia": "PGA.org Website",
  "baseCompensation": 10,
  "baseCompensationMax": null,
  "benefits": null,
  "contact": Object {
    "email": null,
    "name": null,
    "title": null,
  },
  "createdAt": "2018-07-13T14:58:19.742Z",
  "currentStep": "Removed",
  "education": null,
  "employmentType": "Seasonal",
  "experienceRequired": null,
  "hiringManager": Object {
    "employeeId": "139041117",
    "firstName": "Kid",
    "id": 776851,
    "lastName": "Kane",
    "name": undefined,
  },
  "id": 100001,
  "jobCategory": null,
  "jobTitle": "Roman Test Job",
  "offices": Array [
    Object {
      "id": 27259,
      "name": "PGA of America Nat'l Office Headquarters",
    },
  ],
  "openedAt": "2018-07-13T14:58:20.674Z",
  "openings": Array [
    Object {
      "id": 1931104,
      "openedAt": "2018-07-13T14:58:20.299Z",
      "status": "open",
    },
  ],
  "overview": null,
  "paySchedule": "Hourly",
  "responsibilities": null,
  "status": "open",
  "submissionDeadline": "2018-07-14",
  "termOfServiceAgreement": true,
  "totalCompensation": null,
  "userId": undefined,
  "website": null,
  "yearsExperience": null,
}
`;

exports[`Job Model can find user by email 1`] = `
Object {
  "created_at": "2019-02-28T17:12:22.818Z",
  "disabled": false,
  "emails": Array [
    "test@email.com",
  ],
  "employee_id": null,
  "first_name": "first_test",
  "id": 123456,
  "last_name": "last_test",
  "name": "test",
  "site_admin": false,
  "updated_at": "2019-02-28T17:12:22.818Z",
}
`;

exports[`Job Model can get custom fields 1`] = `
Array [
  Object {
    "active": true,
    "customFieldOptions": Array [
      Object {
        "id": 10684619,
        "name": "Seasonal",
        "priority": 0,
      },
      Object {
        "id": 11784837,
        "name": "Year Round",
        "priority": 1,
      },
      Object {
        "id": 11220411,
        "name": "Part Time",
        "priority": 2,
      },
    ],
    "fieldType": "job",
    "id": 131029,
    "name": "Employment Type",
    "nameKey": "employment_type",
    "priority": 2,
    "private": false,
    "requireApproval": false,
    "required": true,
    "triggerNewVersion": false,
    "valueType": "single_select",
  },
]
`;

exports[`Job Model can get job permissions 1`] = `
Array [
  Object {
    "departmentId": null,
    "id": 361480,
    "officeId": 27259,
    "userRoleId": 17735,
  },
]
`;

exports[`Job Model can save a new job 1`] = `
Object {
  "closed_at": null,
  "confidential": false,
  "created_at": "2018-07-13T14:58:19.742Z",
  "custom_fields": Object {
    "additional_income": null,
    "applicant_group": "PGA Professional and General Public",
    "apply_via": "PGA.org Website",
    "base_compensation": Object {
      "unit": "USD",
      "value": 10,
    },
    "base_compensation_max": null,
    "benefits": null,
    "contact_email": null,
    "contact_name": null,
    "contact_title": null,
    "current_step": "Removed",
    "education": null,
    "employment_end_date": null,
    "employment_start_date": null,
    "employment_type": "Seasonal",
    "essential_responsibilities": null,
    "experience_required": null,
    "external_job_url": null,
    "image_set_id": null,
    "job_categories": null,
    "other_job_type": null,
    "overview": null,
    "pay_schedule": "Hourly",
    "submission_deadline": "2018-07-14",
    "supervisor_title": null,
    "term_of_service_agreement": true,
    "total_anticipated_compensation": null,
    "years_experience": null,
  },
  "departments": Array [
    null,
  ],
  "hiring_team": Object {
    "coordinators": Array [],
    "hiring_managers": Array [
      Object {
        "employee_id": "139041117",
        "first_name": "Kid",
        "id": 776851,
        "last_name": "Kane",
      },
    ],
    "recruiters": Array [],
    "sourcers": Array [],
  },
  "id": 100001,
  "name": "Roman Test Job",
  "notes": "Looking for the best!",
  "offices": Array [
    Object {
      "id": 27259,
      "name": "PGA of America Nat'l Office Headquarters",
    },
  ],
  "opened_at": "2018-07-13T14:58:20.674Z",
  "openings": Array [
    Object {
      "application_id": null,
      "close_reason": null,
      "closed_at": null,
      "id": 1931104,
      "opened_at": "2018-07-13T14:58:20.299Z",
      "opening_id": null,
      "status": "open",
    },
  ],
  "requisition_id": null,
  "status": "open",
}
`;

exports[`Job Model can update job 1`] = `
Object {
  "additionalIncome": null,
  "applicantGroup": "PGA Professional and General Public",
  "applyVia": "PGA.org Website",
  "baseCompensation": 10,
  "baseCompensationMax": null,
  "benefits": null,
  "contact": Object {
    "email": null,
    "name": null,
    "title": null,
  },
  "createdAt": "2018-07-13T14:58:19.742Z",
  "currentStep": "Removed",
  "education": null,
  "employmentType": "Seasonal",
  "experienceRequired": null,
  "hiringManager": Object {
    "employeeId": "139041117",
    "firstName": "Kid",
    "id": 776851,
    "lastName": "Kane",
    "name": undefined,
  },
  "id": 100001,
  "jobCategory": null,
  "jobTitle": "Roman Test Job",
  "offices": Array [
    Object {
      "id": 27259,
      "name": "PGA of America Nat'l Office Headquarters",
    },
  ],
  "openedAt": "2018-07-13T14:58:20.674Z",
  "openings": Array [
    Object {
      "id": 1931104,
      "openedAt": "2018-07-13T14:58:20.299Z",
      "status": "open",
    },
  ],
  "overview": null,
  "paySchedule": "Hourly",
  "responsibilities": null,
  "status": "open",
  "submissionDeadline": "2018-07-14",
  "termOfServiceAgreement": true,
  "totalCompensation": null,
  "userId": undefined,
  "website": null,
  "yearsExperience": null,
}
`;

exports[`Job Model getJobsPostedBy 1`] = `
Array [
  Object {
    "additionalIncome": null,
    "applicantGroup": "",
    "applyVia": null,
    "baseCompensation": 14,
    "baseCompensationMax": 55,
    "benefits": "free rounds",
    "contact": Object {
      "email": "jnewkirk@pgahq.com",
      "name": "John Newkirk",
      "title": "PGA Pro",
    },
    "createdAt": "2018-06-14T21:41:58.000Z",
    "currentStep": "Completed",
    "education": Array [
      "bachelor's",
    ],
    "employmentType": "Hourly",
    "experienceRequired": "2 years",
    "hiringManager": Object {
      "employeeId": null,
      "firstName": null,
      "id": null,
      "lastName": null,
    },
    "id": 334,
    "jobCategory": "Seasonal",
    "jobTitle": "PGA Pro",
    "offices": Array [
      Object {
        "city": undefined,
        "id": 123,
        "name": "PGAHQ",
        "state": undefined,
        "zip": undefined,
      },
    ],
    "openedAt": "2018-06-14T21:41:58.000Z",
    "overview": "",
    "paySchedule": "",
    "responsibilities": "Golf professional",
    "status": "open",
    "submissionDeadline": "2018-06-28",
    "termOfServiceAgreement": false,
    "totalCompensation": 66,
    "website": "http://website.com/",
    "yearsExperience": "2 years",
  },
]
`;
